state/unitstate.go

```diff
@@ -0,0 +1,243 @@
+	UniterState string `bson:"uniter-state,omitempty"`
+
+	// RelationState is a serialized yaml string containing relation internal
+	// state for this unit from the uniter.
+	RelationState map[string]string `bson:"relation-state,omitempty"`
+
+	// StorageState is a serialized yaml string containing storage internal
+	// state for this unit from the uniter.
+	StorageState string `bson:"storage-state,omitempty"`
+}
+
+// stateMatches returns true if the State map within the unitStateDoc matches
+// the provided st argument.
+func (d *unitStateDoc) stateMatches(st bson.M) bool {
+	if len(st) != len(d.State) {
+		return false
+	}
+
+	for k, v := range d.State {
+		if st[k] != v {
+			return false
+		}
+	}
+
+	return true
+}
+
+// relationData translate the unitStateDoc's RelationState as
+// a map[string]string to a map[int]string, as is needed.
+// BSON does not allow ints as a map key.
+func (d *unitStateDoc) relationData() (map[int]string, error) {
+	if d.RelationState == nil {
+		return nil, nil
+	}
+	// BSON maps cannot have an int as key.
+	rState := make(map[int]string, len(d.RelationState))
+	for k, v := range d.RelationState {
+		kString, err := strconv.Atoi(k)
+		if err != nil {
+			return nil, err
+		}
+		rState[kString] = v
+	}
+	return rState, nil
+}
+
+// relationStateMatches returns true if the RelationState map within the
+// unitStateDoc matches the provided newRS argument.  Assumes that
+// IgnoreRelationsState has been called first, therefore if arg is empty,
+// returns a nil map to be set.
+func (d *unitStateDoc) relationStateMatches(newRS map[string]string) bool {
+	for k, v := range d.RelationState {
+		if newRS[k] != v {
+			return false
+		}
+	}
+	return true
+}
+
+// removeUnitStateOp returns the operation needed to remove the unit state
+// document associated with the given globalKey.
+func removeUnitStateOp(mb modelBackend, globalKey string) txn.Op {
+	return txn.Op{
+		C:      unitStatesC,
+		Id:     mb.docID(globalKey),
+		Remove: true,
+	}
+}
+
+// UnitState contains the various state saved for this unit,
+// including from the charm itself and the uniter.
+type UnitState struct {
+	// state encodes the unit's persisted state as a list of key-value pairs.
+	state    map[string]string
+	stateSet bool
+
+	// uniterState is a serialized yaml string containing the uniters internal
+	// state for this unit.
+	uniterState    string
+	uniterStateSet bool
+
+	// relationState is a serialized yaml string containing relation internal
+	// state for this unit from the uniter.
+	relationState    map[int]string
+	relationStateSet bool
+
+	// storageState is a serialized yaml string containing storage internal
+	// state for this unit from the uniter.
+	storageState    string
+	storageStateSet bool
+}
+
+// NewUnitState returns a new UnitState struct.
+func NewUnitState() *UnitState {
+	return &UnitState{}
+}
+
+// NoChanges returns true if no values of the struct have been set.
+func (u *UnitState) NoChanges() bool {
```

Comment:
Perhaps `Modified` with an inverted check would be a better name for this method.

Reply:
Is there a better name without an inverted check?

Reply:
Personally I prefer `Modified` or `Mutated` because I like the way the check reads `if !modified { return }`

Reply:
done